### PR TITLE
Added Ledoit Wolf shrinkage covariance estimator

### PR DESCRIPTION
**Pull request by norm4nn**
Added Ledoit Wolf shrinkage covariance matrix estimator. 

[Scikit-learn docs](https://scikit-learn.org/1.5/modules/generated/sklearn.covariance.LedoitWolf.html)

### CODE REVIEW

**Review by josevalim — APPROVED** [review]
LGTM! We only need some tests :)

**Comment by norm4nn** [comment]
I added some tests ;)

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+      end
+
+    location =
+      if opts[:assume_centered] do
+        Nx.broadcast(0, {Nx.axis_size(x, 1)})
+      else
+        Nx.mean(x, axes: [0])
+      end
+
+    {x - Nx.broadcast(location, x), location}
```
Comment (polvalente) [inline lib/scholar/covariance/ledoit_wolf.ex:145]:
```suggestion
    location =
      if opts[:assume_centered] do
        0
      else
        Nx.mean(x, axes: [0])
      end

    {x - location, location}
```

Comment (polvalente) [reply lib/scholar/covariance/ledoit_wolf.ex:145]: I believe the implicit broadcasting should work here

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+
+  defnp ledoit_wolf(x, opts) do
+    case Nx.shape(x) do
+      {_n, 1} ->
+        {
+          Nx.pow(x, 2)
+          |> Nx.mean()
+          |> Nx.broadcast({1, 1}),
+          0.0
+        }
```
Comment (polvalente) [inline lib/scholar/covariance/ledoit_wolf.ex:156]:
```suggestion
        {Nx.mean(x ** 2) |> Nx.reshape({1, 1}), 0.0}
```

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+      _ ->
+        ledoit_wolf_complex(x, opts)
+    end
+  end
+
+  defnp ledoit_wolf_complex(x, opts) do
+    n_features = Nx.axis_size(x, 1)
+    shrinkage = ledoit_wolf_shrinkage(x, opts)
+    emp_cov = empirical_covariance(x, opts)
+    # git
```
Comment (polvalente) [inline lib/scholar/covariance/ledoit_wolf.ex:167]:
```suggestion
```

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+    {shrunk_cov, shrinkage}
+  end
+
+  defnp empirical_covariance(x, _opts) do
+    n = Nx.axis_size(x, 0)
+
+    covariance =
+      Nx.transpose(x)
+      |> Nx.dot(x)
+      |> Nx.divide(n)
```
Comment (polvalente) [inline lib/scholar/covariance/ledoit_wolf.ex:182]:
```suggestion
    covariance = Nx.dot(x, [0], x, [0]) / n
```

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+      _ -> covariance
+    end
+  end
+
+  defnp trace(x) do
+    n = Nx.axis_size(x, 0)
+
+    Nx.eye(n)
+    |> Nx.multiply(x)
+    |> Nx.sum()
```
Comment (polvalente) [inline lib/scholar/covariance/ledoit_wolf.ex:195]:
```suggestion
    x
    |> Nx.take_diagonal()
    |> Nx.sum()
```

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+  end
+
+  defnp ledoit_wolf_shrinkage(x, opts) do
+    case Nx.shape(x) do
+      {_, 1} ->
+        0
+
+      {n} ->
+        Nx.broadcast(x, {1, n})
```
Comment (polvalente) [inline lib/scholar/covariance/ledoit_wolf.ex:204]:
```suggestion
        Nx.reshape(x, {1, n})
```

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+    {n_samples, n_features} = Nx.shape(x)
+    {_, size} = Nx.type(x)
+    block_size = opts[:block_size]
+    n_splits = (n_features / block_size) |> Nx.floor() |> Nx.as_type({:s, size})
+
+    x2 = Nx.pow(x, 2)
+    emp_cov_trace = Nx.sum(x2, axes: [0]) / n_samples
+    mu = Nx.sum(emp_cov_trace) / n_features
+    beta = Nx.tensor(0.0, type: {:f, size})
+    delta = Nx.tensor(0.0, type: {:f, size})
```
Comment (polvalente) [inline lib/scholar/covariance/ledoit_wolf.ex:222]:
```suggestion
    delta = beta = Nx.tensor(0.0, type: {:f, size})
```

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+              {block_size} = Nx.shape(block)
+              rows_from = block_size * i
+              cols_from = block_size * j
+              x2_t = Nx.transpose(x2)
+              x_t = Nx.transpose(x)
+
+              to_beta =
+                Nx.slice_along_axis(x2_t, rows_from, block_size, axis: 0)
+                |> Nx.dot(Nx.slice_along_axis(x2, cols_from, block_size, axis: 1))
+                |> Nx.sum()
```
Comment (polvalente) [inline lib/scholar/covariance/ledoit_wolf.ex:243]:
```suggestion
              to_beta =
                Nx.slice_along_axis(x2, rows_from, block_size, axis: 1)
                |> Nx.dot([0], Nx.slice_along_axis(x2, cols_from, block_size, axis: 1), [0])
                |> Nx.sum()
```

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+
+              beta = beta + to_beta
+
+              to_delta =
+                Nx.slice_along_axis(x_t, rows_from, block_size, axis: 0)
+                |> Nx.dot(Nx.slice_along_axis(x, cols_from, block_size, axis: 1))
+                |> Nx.pow(2)
+                |> Nx.sum()
```
Comment (polvalente) [inline lib/scholar/covariance/ledoit_wolf.ex:251]:
```suggestion
                Nx.slice_along_axis(x, rows_from, block_size, axis: 1)
                |> Nx.dot([0], Nx.slice_along_axis(x, cols_from, block_size, axis: 1), [0])
                |> Nx.pow(2)
                |> Nx.sum()
```

Comment (polvalente) [reply lib/scholar/covariance/ledoit_wolf.ex:251]: this Nx.pow(2) here seems odd, mathematically speaking. is it correct?

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+            |> Nx.dot(Nx.multiply(x2, mask))
+            |> Nx.sum()
+
+          beta = beta + to_beta
+
+          to_delta =
+            Nx.slice_along_axis(x_t, rows_from, block_size, axis: 0)
+            |> Nx.dot(Nx.multiply(x, mask))
+            |> Nx.pow(2)
+            |> Nx.sum()
```
Comment (polvalente) [inline lib/scholar/covariance/ledoit_wolf.ex:276]:
```suggestion
          {m, n} = Nx.shape(x2)
          mask = Nx.iota({m, n}, axis: 1) >= block_size * n_splits

          to_beta =
            Nx.slice_along_axis(x2, rows_from, block_size, axis: 1)
            |> Nx.dot([0], Nx.multiply(x2, mask), [0])
            |> Nx.sum()

          beta = beta + to_beta

          to_delta =
            Nx.slice_along_axis(x, rows_from, block_size, axis: 1)
            |> Nx.dot([0], Nx.multiply(x, mask), [0])
            |> Nx.pow(2)
            |> Nx.sum()
```

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+            |> Nx.dot(Nx.slice_along_axis(x2, cols_from, block_size, axis: 1))
+            |> Nx.sum()
+
+          beta = beta + to_beta
+
+          to_delta =
+            Nx.multiply(x_t, mask)
+            |> Nx.dot(Nx.slice_along_axis(x, cols_from, block_size, axis: 1))
+            |> Nx.pow(2)
+            |> Nx.sum()
```
Comment (polvalente) [inline lib/scholar/covariance/ledoit_wolf.ex:309]:
```suggestion
          {rows, cols} = Nx.shape(x)

          mask = Nx.iota({cols, rows}, axis: 0) >= block_size * n_splits

          to_beta =
            Nx.multiply(x2, mask)
            |> Nx.dot([0], Nx.slice_along_axis(x2, cols_from, block_size, axis: 1), [0])
            |> Nx.sum()

          beta = beta + to_beta

          to_delta =
            Nx.multiply(x, mask)
            |> Nx.dot([0], Nx.slice_along_axis(x, cols_from, block_size, axis: 1), [0])
            |> Nx.pow(2)
            |> Nx.sum()
```

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+      |> Nx.pow(2)
+      |> Nx.sum()
+
+    delta = delta + to_delta
+    delta = delta / Nx.pow(n_samples, 2)
+
+    to_beta =
+      Nx.multiply(x2_t, mask_t)
+      |> Nx.dot(Nx.multiply(x2, mask))
+      |> Nx.sum()
```
Comment (polvalente) [inline lib/scholar/covariance/ledoit_wolf.ex:339]:
```suggestion
    {rows, cols} = Nx.shape(x)
    mask = Nx.iota({rows, cols}, axis: 1) >= block_size * n_splits
    
    to_delta =
      Nx.multiply(x, mask)
      |> Nx.dot([0], Nx.multiply(x, mask), [0])
      |> Nx.pow(2)
      |> Nx.sum()

    delta = delta + to_delta
    delta = delta / Nx.pow(n_samples, 2)

    to_beta =
      Nx.multiply(x2, mask)
      |> Nx.dot([0], Nx.multiply(x2, mask), [0])
      |> Nx.sum()
```

File: test/scholar/covariance/ledoit_wolf.exs [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        Nx.tensor([0.0, 0.0]),
+        Nx.tensor([[2.2, 1.5], [0.7, 1.1]]),
+        shape: {50},
+        type: :f32
+      )
+
+    x
+  end
+
+  defp x3 do
```
Comment (polvalente) [inline test/scholar/covariance/ledoit_wolf.exs:40]: I think these shouldn't be defined separately from the tests that use them. I'm fine with `key` being a defp though

File: test/scholar/covariance/ledoit_wolf.exs [hunk 14]
```diff
@@ -0,0 +1,126 @@
+defmodule Scholar.Covariance.LedoitWolfTest do
```
Comment (polvalente) [inline test/scholar/covariance/ledoit_wolf.exs:1]: I'm not sure if this test is actually running. It's missing the `_test` suffix in the filename

**Comment by msluszniak** [comment]
blocks like 
```elixir
      Nx.multiply(x, mask)
      |> Nx.dot([0], Nx.multiply(x, mask), [0])
      |> Nx.pow(2)
      |> Nx.sum()
```
occur multiple times. You might create a function simplify this

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+      
+      iex> key = Nx.Random.key(0)
+      iex> {x, _new_key} = Nx.Random.multivariate_normal(key, Nx.tensor([0.0, 0.0, 0.0]), Nx.tensor([[3.0, 2.0, 1.0], [1.0, 2.0, 3.0], [1.3, 1.0, 2.2]]), shape: {10}, type: :f32)
+      iex> model = Scholar.Covariance.LedoitWolf.fit(x)
+      iex> model.covariance
+      #Nx.Tensor<
+        f32[3][3]
+        [
+          [2.5945029258728027, 1.507835865020752, 1.1623677015304565],
+          [1.507835865020752, 2.106797218322754, 1.181215524673462],
```
[the remark this answers is not in the document]
Comment (norm4nn) [reply lib/scholar/covariance/ledoit_wolf.ex:84]: It is, at least I'm getting same results calling it scikit-learn version of Ledoit-Wolf:
```
import numpy as np
from sklearn.covariance import LedoitWolf

X = np.array(
    [
     [3.4786570072174072, 1.6614335775375366, 1.7482566833496094],
     [0.07869422435760498, -1.9089748859405518, -0.3356112539768219],
     [-1.475188970565796, -0.6677497029304504, -1.024237871170044],
     [0.37418487668037415, 1.0414379835128784, 1.676888346672058],
     [2.0346086025238037, 0.9449125528335571, 1.9334920644760132],
     [3.13038969039917, 1.7878211736679077, 0.888858437538147],
     [0.7129192352294922, 1.125357985496521, 1.627787709236145],
     [1.7261531352996826, 2.394443988800049, 1.8893500566482544],
     [-1.3257558345794678, -2.0037055015563965, -1.072041392326355],
     [2.4940638542175293, 1.0443227291107178, 1.3461097478866577]
   ],
  dtype=np.float64)

cov = LedoitWolf().fit(X)

(cov.covariance_, cov.shrinkage_, cov.location_)
```

Yeah, I guess it's not the best example to show the usage of this algo. Let me know if you want me to change it.

[the remark this answers is not in the document]
Comment (polvalente) [reply lib/scholar/covariance/ledoit_wolf.ex:84]: No need to change!

File: lib/scholar/covariance/ledoit_wolf.ex (hunk 9 above)
Comment (norm4nn) [reply lib/scholar/covariance/ledoit_wolf.ex:248]: I’ll double-check that, but at first glance - it looks correct.

Comment (norm4nn) [reply lib/scholar/covariance/ledoit_wolf.ex:248]: So, if we take the [scikit-learn implementation](https://github.com/scikit-learn/scikit-learn/blob/6e9039160f0dfc3153643143af4cfdca941d2045/sklearn/covariance/_shrunk_covariance.py#L376) as a reference, then it is correct. But I have to agree with you - this looks weird. I don't see a reason to implement it that way - I will simplify it in the next commit.

Comment (norm4nn) [reply lib/scholar/covariance/ledoit_wolf.ex:248]: I will simplify this whole section significantly in the next commit. After longer investigation of the scikit-learn code, I found it highly overcomplicated for Scholar. Maybe in Python it has some reason to calculate beta and delta in the loop, but I don't believe this applies here.

**Comment by norm4nn** [comment]
Sorry for changing a lot of code in the middle of review, but I believe that now it's much more readable

**Review by josevalim — APPROVED** [review]

**Review by msluszniak — APPROVED** [review]
Lgtm :))

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+    case Nx.shape(x) do
+      {_n, 1} ->
+        {Nx.mean(x ** 2) |> Nx.reshape({1, 1}), 0.0}
+
+      _ ->
+        ledoit_wolf_shrinkage(x)
+    end
+  end
+
+  defnp empirical_covariance(x) do
```
Comment (krstopro) [inline lib/scholar/covariance/ledoit_wolf.ex:152]: I think this can be replaced with `Nx.covariance/3`.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

File: lib/scholar/covariance/ledoit_wolf.ex [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+  Covariance Matrices", Ledoit and Wolf, Journal of Multivariate
+  Analysis, Volume 88, Issue 2, February 2004, pages 365-411.
+  """
+  import Nx.Defn
+
+  @derive {Nx.Container, containers: [:covariance, :shrinkage, :location]}
+  defstruct [:covariance, :shrinkage, :location]
+
+  opts_schema = [
+    assume_centered: [
```
Comment (krstopro) [inline lib/scholar/covariance/ledoit_wolf.ex:16]: I would rename it to `assume_centered?` (add `?`).

**Review by krstopro — COMMENTED** [review]
Two comments from my side.

**Comment by msluszniak** [comment]
@krstopro if `Nx.covariance` works, then feel free to merge these changes in seperate pr/commit

**Comment by krstopro** [comment]
> @krstopro if `Nx.covariance` works, then feel free to merge these changes in seperate pr/commit

Sure, but lemme fix #301 first. :)

**Comment by josevalim** [comment]
@norm4nn can also submit a pull request with the Nx.covariance fix if they have the time :)

**Pull request merged**